libraries/chain/db_block.cpp

```diff
@@ -626,9 +626,12 @@ processed_transaction database::_apply_transaction(const signed_transaction& trx
 
    auto& trx_idx = get_mutable_index_type<transaction_index>();
    const chain_id_type& chain_id = get_chain_id();
-   auto trx_id = trx.id();
-   FC_ASSERT( (skip & skip_transaction_dupe_check) ||
-              trx_idx.indices().get<by_trx_id>().find(trx_id) == trx_idx.indices().get<by_trx_id>().end() );
+   transaction_id_type trx_id;
```

Comment:
I like that you did this. It is more easily readable IMO. Currently, you're protected with the if. But hopefully someone later will know better to not trust that trx_id may not be set.

Reply:
Good point, will address this in #1360 (id is cached inside transaction there)

Reply:
To avoid using an uninitialized id, perhaps wrap it with `optional`? May cause some overheads though.